File: 2015/14/solution.js
let reindeer = {}
let distances = {}
let scores = {}

const maximum = (obj) => Math.max.apply(null, Object.values(obj))

const defineDeer = (line) => {
  const matcher = /^(.*) can fly (\d+) km\/s for (\d+) seconds, but then must rest for (\d+) seconds\.$/
  const [ , deer, speed, flyTime, restTime ] = line.match(matcher)
    reindeer[deer] = {
    speed: parseInt(speed),
    flyTime: parseInt(flyTime),
    restTime: parseInt(restTime),
    flying: true,
    timeInState: 0
  }
}

const iterateDeer = (func) => Object.keys(reindeer).forEach(func)

const deerDoneCurrentState = (deer) => {
  return deer.timeInState === deer[deer.flying ? 'flyTime' : 'restTime']
}

const incrementDeerTime = (deer) => {
  deer.timeInState += 1
  if (deerDoneCurrentState(deer)) {
    deer.flying = !deer.flying
    deer.timeInState = 0
  }
}

const tick = () => {
  iterateDeer((deer) => {
    const { speed, flying } = reindeer[deer]
    if (flying) distances[deer] = (distances[deer] || 0) + speed
    incrementDeerTime(reindeer[deer])
  })
  const currentMaximum = maximum(distances)
  iterateDeer((deer) => {
    if ((distances[deer] || 0) === currentMaximum) {
      scores[deer] = (scores[deer] || 0) + 1
    }
  })
}

const report = () => {
  for (let i = 0; i < 2503; i++) tick()
  console.log('Part 1', maximum(distances))
  console.log('Part 2', maximum(scores))
}

require('readline').
    createInterface({ input: require('fs').createReadStream('input.txt') }).
    on('line', defineDeer).
    on('close', report)

File: 2015/14/solution.rb
class ReindeerRace

  attr_accessor :reindeer, :distances, :scores

  def initialize(file)
    @reindeer = {}
    @distances = Hash.new(0)
    @scores = Hash.new(0)

    file.each_line do |line|
      matcher = /^(.*) can fly (\d+) km\/s for (\d+) seconds, but then must rest for (\d+) seconds\.$/
      deer, speed, fly_time, rest_time = matcher.match(line)[1..4]
      @reindeer[deer] = {
        speed: speed.to_i,
        fly_time: fly_time.to_i,
        rest_time: rest_time.to_i,
        flying: true,
        time_in_state: 0
      }
    end
  end

  def tick
    reindeer.keys.each do |deer|
      if reindeer[deer][:flying]
        distances[deer] += reindeer[deer][:speed]
      end
      increment_deer_time(reindeer[deer])
    end
    current_max = distances.values.max
    reindeer.keys.each do |deer|
      if distances[deer] == current_max
        scores[deer] += 1
      end
    end
  end

  def max_distance
    distances.values.max
  end

  def max_score
    scores.values.max
  end

  protected
  def deer_done_current_state(deer)
    deer[:time_in_state] == deer[deer[:flying] ? :fly_time : :rest_time]
  end

  def increment_deer_time(deer)
    deer[:time_in_state] += 1
    if deer_done_current_state(deer)
      deer[:flying] = !deer[:flying]
      deer[:time_in_state] = 0
    end
  end

end

race = ReindeerRace.new(File.open('input.txt'))

2503.times { race.tick }

puts 'Part 1', race.max_distance
puts 'Part 2', race.max_score
